### PR TITLE
Add sets

### LINKED ISSUE #83: [FEAT] Set type
## Describe your feature
A set type would allow for sending dictionaries where all the values are `true`. This can be optimized over a map since only the keys have to be sent.

## Alternatives
Literal types like `true` or `false` could be added. Then a `map { [string]: true }` could be used instead assuming zap would be smart enough to optimize for this.

### PR DESCRIPTION
**Pull request by Ketasaja**
Closes #83.

### CODE REVIEW

File: zap/src/irgen/ser.rs [hunk 1]
```diff
@@ -201,6 +201,34 @@ impl Ser {
...
 
+			Ty::Set(key) => {
+				let (len_name, len_expr) = self.add_occurrence("len");
+				let (len_pos_name, len_pos_expr) = self.add_occurrence("len_pos");
+
+				self.push_local(len_pos_name.clone(), Some(Var::from("alloc").call(vec![2.0.into()])));
+				self.push_local(len_name.clone(), Some(0.0.into()));
+
+				let (key_name, _) = self.add_occurrence("k");
+				let (val_name, _) = self.add_occurrence("_");
```
Comment (sasial-dev) [inline zap/src/irgen/ser.rs:212]: Does `_` get picked up for shadowing and need an `add_occurrence`?

File: zap/src/output/typescript/mod.rs [hunk 2]
```diff
@@ -119,6 +119,12 @@ pub trait Output: ConfigProvider {
 				self.push(">");
 			}
 
+			Ty::Set(key) => {
+				self.push("Map<");
```
Comment (sasial-dev) [inline zap/src/output/typescript/mod.rs]:
```suggestion
				self.push("Set<");
```

File: zap/src/output/typescript/mod.rs [hunk 3]
```diff
@@ -119,6 +119,12 @@ pub trait Output: ConfigProvider {
...
+				self.push_ty(key);
+				self.push(", true>");
```
Comment (sasial-dev) [inline zap/src/output/typescript/mod.rs]:
```suggestion
				self.push(">");
```

**Review by sasial-dev — CHANGES_REQUESTED** [review]
Just these minor things otherwise the code LGTM.

File: zap/src/irgen/ser.rs (hunk 1 above)
Comment (sasial-dev) [reply zap/src/irgen/ser.rs:212]: Maybe it is good practice to keep it with `add_occurrence`? Curious to know what you think.

Comment (Ketasaja) [reply zap/src/irgen/ser.rs:212]: It becomes `__1`. Selene doesn't care about shadowing `_` but you're not writing Luau normally so there is no warning if you use an `_` variable. I'm fine with whichever, I could also add string splitting back to emit `_1` if a variable ends in `_`.

Comment (sasial-dev) [reply zap/src/irgen/ser.rs:212]: Let's keep it as is then. No need to make it emit `_1` I reckon.

File: zap/src/irgen/des.rs [hunk 4]
```diff
@@ -187,6 +187,24 @@ impl Des {
...
+
+				self.push_stmt(Stmt::NumFor {
+					var: "_".into(),
+					from: 1.0.into(),
+					to: self.readu16(),
+				});
+
+				let (key_name, key_expr) = self.add_occurrence("key");
+
+				self.push_ty(key, Var::Name(key_name.clone()));
```
Comment (sasial-dev) [inline zap/src/irgen/des.rs]: `key_1` isn't defined first with
`local key_1`, and so the output looks like this:
```lua
			local value
			value = {}
			for _ = 1, buffer.readu16(incoming_buff, read(2)) do
				key_1 = buffer.readu8(incoming_buff, read(1))
				value[key_1] = true
			end
```

**Review by sasial-dev — CHANGES_REQUESTED** [review]

**Review by sasial-dev — APPROVED** [review]
LGTM!

**Pull request merged**